rust/processor/src/worker.rs

```diff
@@ -706,3 +582,168 @@ pub async fn get_stream(
+                    stream_address = indexer_grpc_data_service_address.clone(),
+                    error = ?rpc_error,
+                    "[Parser] Error receiving datastream response."
+                );
+                false
+            },
+            None => {
+                tracing::warn!(
+                    processor_name = processor_name,
+                    stream_address = indexer_grpc_data_service_address.clone(),
+                    "[Parser] Stream ended."
+                );
+                false
+            },
+        };
+        // Check if we're at the end of the stream
+        let is_end = if let Some(ending_version) = request_ending_version {
+            next_version_to_fetch > ending_version
+        } else {
+            false
+        };
+        if is_end {
+            info!(
+                processor_name = processor_name,
+                stream_address = indexer_grpc_data_service_address.clone(),
+                ending_version = request_ending_version,
+                next_version_to_fetch = next_version_to_fetch,
+                "[Parser] Reached ending version.",
+            );
+            // Wait for the fetched transactions to finish processing before closing the channel
+            loop {
+                let channel_capacity = tx.capacity();
+                info!(
+                    processor_name = processor_name,
+                    channel_size = BUFFER_SIZE - channel_capacity,
+                    "[Parser] Waiting for channel to be empty"
+                );
+                if channel_capacity == BUFFER_SIZE {
+                    break;
+                }
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+            info!("[Parser] The stream is ended.");
+        } else {
+            // The rest is to see if we need to reconnect
+            if is_success {
+                continue;
+            }
+            if let Some(lrt) = last_reconnection_time {
+                let elapsed: u64 = lrt.elapsed().as_secs();
+                if reconnection_retries >= RECONNECTION_MAX_RETRIES
+                    && elapsed < MIN_SEC_BETWEEN_GRPC_RECONNECTS
+                {
+                    error!(
+                        processor_name = processor_name,
+                        stream_address = indexer_grpc_data_service_address.clone(),
+                        seconds_since_last_retry = elapsed,
+                        "[Parser] Recently reconnected. Will not retry.",
+                    );
+                    panic!("[Parser] Recently reconnected. Will not retry.")
+                }
+            }
+            reconnection_retries += 1;
+            last_reconnection_time = Some(std::time::Instant::now());
+            tracing::warn!(
+                processor_name = processor_name,
+                stream_address = indexer_grpc_data_service_address.clone(),
+                starting_version = next_version_to_fetch,
```

Comment:
If we panic early (before the consumer channel finishes processing, using `next_version_to_fetch` as the new starting version would potentially skip some transactions that were lost in the consumer channel. 

Which is why I suggested waiting for the consumer channel to clear before attempting reconnect.

Reply:
wait what how is this possible?